keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,524 @@
+| `duplicatemarkers` | All | Checks that fields in the API types do not contain duplicate markers. | N/A | No |
+| `enums` | All | Enumerations should use a type alias and `+enum` marker. Enum values should be PascalCase | Pattern to match for Enum values, a list of exceptions to allow (e.g. for CLI tool names) | No |
+| `nameformats` | CRD (future All) | Checks that common fields like `Name` and `Namespace` in API types are formatted correctly, using CEL and the format library. | N/A | No |
+| `noduration` | All | Checks that fields in the API types do not contain a `time.Duration` type. Should be `fooSeconds` or `fooMinutes` etc. | N/A | No |
+| `nomaps` | All | Checks that fields in the API types do not contain a `map` type with a struct value. Maps of subobjects are not allowed. | N/A | No |
+| `numericbounds` | CRD (future All) | Checks that numeric fields in the API types are bounded by a maximum and minimum value. | N/A | No |
+| `optionalfields` | All | Checks that all fields in an API type marked as optional follow the convention of being pointers and having `omitempty` in their JSON tag. | Option to allow non-pointers for non-struct types. | No |
+| `references` | All | Checks that fields are named `Ref/Refs` and not `Reference/References`. | Option to forbid usage of either `Ref` or `Reference`. | No |
+| `ssatags` | All | Checks that arrays have a `listType` marker | Option to forbid the use of `listType=set` for array of struct | No |
+| `statusoptional` | All | Checks that all first level children of a `status` field are marked as optional. | N/A | No |
+| `timestamps` | All | Checks that fields in the API types that are timestamps are named `FooTime` and not `FooTimestamp`. | N/A | No |
+| `typeandobjectmeta` | All | Checks for top-level types (i.e. Kinds) and checks that they have inline typemeta and objectmata is optional/omitempty. | N/A | No |
+
+A number of rules implemented are currently only applicable to CRDs, but in the future could be applicable to built-in types as well.
+There is on-going research into marking built-ins and generating validation from these markers.
+Once this research reaches its conclusion, these markers could be integrated into the rules that are at present, CRD only, to make them applicable to all types.
+
+The unimplemented rules are being tracked by an [open issue](https://github.com/JoelSpeed/kal/issues/1) on the repository.
+This enhancement is not aimed at discussion about individual rules, and comments about the rules may be left on their individual tracking issues.
+
+### CRD Schema Change Validation
+
+There is some previous work done in both https://github.com/everettraven/crd-diff and https://github.com/openshift/crd-schema-checker. Each project implements schema checks in different ways and has different user workflows.
+
+It is proposed that a new repository is created that merges select bits and pieces from each project into a single project.
+
+Currently, it is proposed that:
```

Comment:
Shared unit tests for admission and CLI validator is important for compatibility/consistency.  crd-schema-checker does this.